utils/managed_bytes.hh

```diff
@@ -104,41 +139,82 @@ struct [[gnu::packed]] medium_blob_storage {
 };
 
 // A managed version of "bytes" (can be used with LSA).
+//
+// Sometimes also used as a general-purpose fragmented buffer outside of LSA context,
+// but this is not recommended, because it's too easy to accidentally destroy it
+// in a different allocator than it was allocated in, which can break the program
+// in a hard-to-predict way.
+//
+// managed_bytes has three storage layouts:
+// 1. Inline.
+//    Used for data which fits into max_inline_size.
+// 2. External contiguous. (Single-allocation).
+//    Used for data which fits into preferred_max_contiguous_allocation().
+//    (At the moment of writing: 128 kiB and 12,8 kiB in LSA).
+//    The storage is a single medium_blob_storage object.
+// 3. External fragmented. (Multi-allocation).
+//    Used for everything else.
+//    The storage is a chain of blob_storage objects.
+//
+// Layout 2 exists as an optimization for the most common allocation sizes (several bytes).
+// There is nothing which prevents implementing these with layout 3, but layout 3 stores slightly
+// more metadata in the allocated buffer (pointer to the next fragment, size of the current fragment),
+// which adds up to a big overhead when used with small allocations.
+// E.g. 8-byte values are allocated externally -- each has additional 1 byte of flags and 8 bytes
+// of timestamp, so it's 17 bytes in total and that doesn't fit into inline storage.
+// And adding 16 bytes to each 17-byte cell is a big waste.
+//
+// The code of `class managed_bytes` is responsible for allocating and freeing the storage.
+// Code responsible for reading and writing it is in managed_bytes_basic_view.
+// The implementation details of these two classes are entangled.
 class managed_bytes {
     friend class bytes_ostream;
     static constexpr size_t max_inline_size = 15;
+
+    // The current layout is discerned from the leading `inline_size` of the union.
+    // >0 -> layout 1 (inline).
+    // -1 -> layout 2 (medium_blob_storage)
+    // -2 -> layout 3 (blob_storage)
+    //
+    // Note: reading a union member other than last modified is generally
+    // forbidden by the standard, but reading members of the "initial common subsequence"
+    // of fields is permitted.
+    // (Compilers are generally much more permissive than the standard and allow
+    // full-blown type punning via unions, because the standard rules are neither
+    // intuitive nor useful).
     struct small_blob {
+        int8_t inline_size; // In layout 1, stores the buffer size.
         bytes_view::value_type data[max_inline_size];
-        int8_t size; // -1 -> use blob_storage
-                     // -2 -> use medium_blob_storage
     };
     struct medium_blob {
-        medium_blob_storage::ref_type ref;
-        int8_t padding[4];
+        int8_t inline_size; // In layout 2, always equals -1.
+        int8_t padding[3]; // Only used to pad `ref` to its natural alignment. (Could help with codegen).
```

Comment:
7 would be better, perhaps another patch.

Reply:
No, because the medium_blob_storage::ref_type is 12 bytes, not 8 bytes (it contains not just the pointer, but also the size.

It can't be hoisted outside because the medium_blob_storage has to be able to find out its size during LSA compaction.